File: python/fastscapelib/tests/test_flow_graph.py
from textwrap import dedent

import numpy as np
import numpy.testing as npt
import pytest

from fastscapelib.flow import (
    FlowGraph,
    FlowSnapshot,
    MultiFlowRouter,
    PFloodSinkResolver,
    SingleFlowRouter,
)
from fastscapelib.grid import NodeStatus, ProfileGrid, RasterGrid


class TestFlowGraph:
    def test___init__(self) -> None:
        profile_grid = ProfileGrid(8, 2.2, NodeStatus.FIXED_VALUE)
        raster_grid = RasterGrid([5, 10], [2.2, 2.4], NodeStatus.FIXED_VALUE)

        FlowGraph(profile_grid, [SingleFlowRouter()])
        FlowGraph(raster_grid, [PFloodSinkResolver(), SingleFlowRouter()])

        with pytest.raises(
            ValueError, match="must have at least one operator that updates"
        ):
            FlowGraph(raster_grid, [])

        with pytest.raises(TypeError, match="invalid flow operator"):
            FlowGraph(raster_grid, ["not a flow operator"])  # type: ignore[list-item]

    def test_base_levels(self) -> None:
        raster_grid = RasterGrid([3, 3], [1.0, 1.0], NodeStatus.FIXED_VALUE)
        flow_graph = FlowGraph(raster_grid, [SingleFlowRouter()])

        actual = flow_graph.base_levels
        assert set(actual) == {0, 1, 2, 3, 5, 6, 7, 8}

        flow_graph.base_levels = [3]
        assert flow_graph.base_levels == [3]

    def test_mask(self) -> None:
        raster_grid = RasterGrid([3, 3], [1.0, 1.0], NodeStatus.FIXED_VALUE)
        flow_graph = FlowGraph(raster_grid, [SingleFlowRouter()])

        # uninitialized mask
        npt.assert_array_equal(flow_graph.mask, np.array([], dtype=np.bool_))

        flow_graph.mask = np.ones(raster_grid.shape, dtype=np.bool_)
        npt.assert_array_equal(
            flow_graph.mask, np.ones(raster_grid.shape, dtype=np.bool_)
        )

    def test_operators(self) -> None:
        grid = ProfileGrid(8, 2.2, NodeStatus.FIXED_VALUE)

        resolver = PFloodSinkResolver()
        router = SingleFlowRouter()

        flow_graph = FlowGraph(grid, [resolver, router])

        assert flow_graph.operators == [resolver, router]

    def test_single_flow(self) -> None:
        grid = ProfileGrid(8, 2.2, NodeStatus.FIXED_VALUE)

        flow_graph = FlowGraph(grid, [SingleFlowRouter()])
        assert flow_graph.single_flow is True

        flow_graph = FlowGraph(grid, [MultiFlowRouter(1.0)])
        assert flow_graph.single_flow is False

    def test_repr(self) -> None:
        grid = ProfileGrid(8, 2.2, NodeStatus.FIXED_VALUE)
        flow_graph = FlowGraph(grid, [SingleFlowRouter(), FlowSnapshot("test")])

        actual = repr(flow_graph)
        expected = """
        <FlowGraph (8 nodes)>
        Operators:
            SingleFlowRouter
            FlowSnapshot 'test' (graph=True, elevation=False)
        """
        assert actual == dedent(expected[1:])

        actual = repr(flow_graph.graph_snapshot("test"))
        expected = """
        <FlowGraph (8 nodes)>
        Operators:
            *empty*
        """
        assert actual == dedent(expected[1:])

    def test_update_routes(self) -> None:
        grid = ProfileGrid(8, 2.2, NodeStatus.FIXED_VALUE)
        flow_graph = FlowGraph(grid, [SingleFlowRouter()])

        # pit at 3rd node
        elevation = np.array([0.0, 0.2, 0.1, 0.2, 0.4, 0.6, 0.3, 0.0])

        new_elevation = flow_graph.update_routes(elevation)

        npt.assert_array_equal(elevation, new_elevation)

        npt.assert_equal(
            flow_graph.impl().receivers[:, 0], np.array([0, 0, 2, 2, 3, 6, 7, 7])
        )
        npt.assert_equal(flow_graph.impl().receivers_count, np.ones(elevation.size))
        npt.assert_equal(
            flow_graph.impl().receivers_weight[:, 0], np.ones(elevation.size)
        )

        actual = flow_graph.impl().donors
        m = np.iinfo(actual.dtype).max
        npt.assert_equal(
            actual,
            np.array(
                [
                    [1, m, m],
                    [m, m, m],
                    [2, 3, m],
                    [4, m, m],
                    [m, m, m],
                    [m, m, m],
                    [5, m, m],
                    [6, m, m],
                ]
            ),
        )
        npt.assert_equal(
            flow_graph.impl().donors_count, np.array([1, 0, 2, 1, 0, 0, 1, 1])
        )

    def test_accumulate_basins(self) -> None:
        # --- test profile grid
        grid = ProfileGrid(8, 2.0, NodeStatus.FIXED_VALUE)
        flow_graph = FlowGraph(grid, [SingleFlowRouter()])

        # pit at 3rd node
        elevation = np.array([0.0, 0.2, 0.1, 0.2, 0.4, 0.6, 0.3, 0.0])

        new_elevation = flow_graph.update_routes(elevation)

        npt.assert_array_equal(elevation, new_elevation)

        acc = np.empty_like(elevation)
        src = np.ones_like(elevation)
        expected = np.array([4.0, 2.0, 6.0, 4.0, 2.0, 2.0, 4.0, 6.0])

        flow_graph.accumulate(acc, 1.0)
        npt.assert_array_equal(acc, expected)

        flow_graph.accumulate(acc, src)
        npt.assert_array_equal(acc, expected)

        npt.assert_almost_equal(flow_graph.accumulate(1.0), expected)
        npt.assert_almost_equal(flow_graph.accumulate(src), expected)

        # bottom border base-level
        bottom_base_level = [
            NodeStatus.CORE,
            NodeStatus.CORE,
            NodeStatus.CORE,
            NodeStatus.FIXED_VALUE,
        ]

        # --- test raster grid
        rgrid = RasterGrid([4, 4], [1.0, 1.0], bottom_base_level)

        flow_graph = FlowGraph(rgrid, [SingleFlowRouter()])

        # planar surface tilted along the y-axis + small carved channel
        elevation = np.array(
            [
                [0.6, 0.6, 0.6, 0.6],
                [0.4, 0.4, 0.4, 0.4],
                [0.2, 0.2, 0.2, 0.2],
                [0.1, 0.0, 0.1, 0.1],
            ]
        )

        new_elevation = flow_graph.update_routes(elevation)
        npt.assert_array_equal(elevation, new_elevation)

        acc = np.empty_like(elevation)
        src = np.ones_like(elevation)
        expected = np.array(
            [
                [1.0, 1.0, 1.0, 1.0],
                [2.0, 2.0, 2.0, 2.0],
                [3.0, 3.0, 3.0, 3.0],
                [1.0, 10.0, 1.0, 4.0],
            ]
        )

        flow_graph.accumulate(acc, 1.0)
        npt.assert_array_equal(acc, expected)

        flow_graph.accumulate(acc, src)
        npt.assert_array_equal(acc, expected)

        npt.assert_almost_equal(flow_graph.accumulate(1.0), expected)
        npt.assert_almost_equal(flow_graph.accumulate(src), expected)

        expected = np.array(
            [
                [1, 1, 1, 3],
                [1, 1, 1, 3],
                [1, 1, 1, 3],
                [0, 1, 2, 3],
            ]
        )
        npt.assert_array_equal(flow_graph.basins(), expected)
        npt.assert_array_equal(flow_graph.basins().flatten(), flow_graph.impl().basins)

        # test basins mask
        mask = np.array(
            [
                [False, False, False, True],
                [False, False, False, True],
                [False, False, False, True],
                [False, False, False, True],
            ]
        )

        flow_graph.mask = mask
        flow_graph.update_routes(elevation)
        actual = flow_graph.basins()
        no_basin = np.iinfo(actual.dtype).max
        expected = np.array(
            [
                [1, 1, 1, no_basin],
                [1, 1, 1, no_basin],
                [1, 1, 1, no_basin],
                [0, 1, 2, no_basin],
            ]
        )
        npt.assert_array_equal(actual, expected)

File: test/test_flow_graph.cpp
#include <algorithm>

#include "xtensor/xtensor.hpp"
#include "xtensor/xstrided_view.hpp"

#include "gtest/gtest.h"

#include "fastscapelib/flow/flow_graph.hpp"
#include "fastscapelib/flow/flow_router.hpp"
#include "fastscapelib/flow/flow_snapshot.hpp"
#include "fastscapelib/grid/raster_grid.hpp"


namespace fs = fastscapelib;

namespace fastscapelib
{
    namespace testing
    {

        class flow_graph : public ::testing::Test
        {
        protected:
            using flow_graph_type = fs::flow_graph<fs::raster_grid>;
            using grid_type = fs::raster_grid;
            using size_type = typename grid_type::size_type;

            // bottom border base-level
            fs::node_status fixed = fs::node_status::fixed_value;
            fs::node_status core = fs::node_status::core;
            fs::raster_boundary_status bottom_base_level{ { core, core, core, fixed } };

            grid_type grid = grid_type({ 4, 4 }, { 1.0, 1.0 }, bottom_base_level);

            // planar surface tilted along the y-axis + small carved channel
            xt::xtensor<double, 2> elevation{ { 0.6, 0.6, 0.6, 0.6 },
                                              { 0.4, 0.4, 0.4, 0.4 },
                                              { 0.2, 0.2, 0.2, 0.2 },
                                              { 0.1, 0.0, 0.1, 0.1 } };
        };

        TEST_F(flow_graph, ctor)
        {
            auto graph = flow_graph_type(grid, { fs::single_flow_router() });

            EXPECT_EQ(graph.size(), 16u);
            EXPECT_TRUE(xt::same_shape(graph.grid_shape(), grid.shape()));

            // no operator updating the graph (flow router)
            EXPECT_THROW(flow_graph_type(grid, { fs::flow_snapshot("s") }), std::invalid_argument);
        }

        TEST_F(flow_graph, base_levels)
        {
            auto graph = flow_graph_type(grid, { fs::single_flow_router() });

            {
                SCOPED_TRACE("default base levels");

                auto actual = graph.base_levels();
                // base_levels are not ordered (implementation detail)
                std::sort(actual.begin(), actual.end());
                EXPECT_EQ(actual, std::vector<size_type>({ 12, 13, 14, 15 }));
            }
            {
                SCOPED_TRACE("set base levels");

                graph.set_base_levels(std::vector<size_type>({ 5 }));
                EXPECT_EQ(graph.base_levels(), std::vector<size_type>({ 5 }));
            }
        }

        TEST_F(flow_graph, mask)
        {
            auto graph = flow_graph_type(grid, { fs::single_flow_router() });

            {
                SCOPED_TRACE("uninitialized mask (default)");

                EXPECT_EQ(graph.mask(), xt::xarray<bool>());
            }
            {
                SCOPED_TRACE("set mask");

                graph.set_mask(xt::ones<bool>(grid.shape()));
                EXPECT_EQ(graph.mask(), xt::ones<bool>(grid.shape()));
            }
            {
                SCOPED_TRACE("set mask error (shape mismatch)");

                EXPECT_THROW(graph.set_mask(xt::ones<bool>({ 10. })), std::runtime_error);
            }
        }

        TEST_F(flow_graph, operators)
        {
            auto graph
                = flow_graph_type(grid, { fs::single_flow_router(), fs::flow_snapshot("s") });

            std::vector<std::string> expected{ "single_flow_router", "flow_snapshot" };
            std::vector<std::string> actual;

            for (const auto* op : graph.operators())
            {
                actual.push_back(op->name());
            }

            EXPECT_EQ(actual, expected);
        }

        TEST_F(flow_graph, single_flow)
        {
            {
                SCOPED_TRACE("single flow is true");

                auto graph = flow_graph_type(grid, { fs::single_flow_router() });
                EXPECT_EQ(graph.single_flow(), true);
            }
            {
                SCOPED_TRACE("single flow is false");

                auto graph = flow_graph_type(grid, { fs::multi_flow_router(1.0) });
                EXPECT_EQ(graph.single_flow(), false);
            }
        }

        TEST_F(flow_graph, update_routes)
        {
            auto graph = flow_graph_type(grid, { fs::single_flow_router() });

            auto new_elevation = graph.update_routes(elevation);

            // re-run to check there is no memory effect
            graph.update_routes(elevation);

            EXPECT_TRUE(xt::all(xt::equal(new_elevation, elevation)));

            // more in-depth testing is done in flow router (methods) tests
            auto actual = xt::col(graph.impl().receivers(), 0);
            xt::xtensor<size_type, 1> expected{ 4,  5,  6,  7,  8,  9,  10, 11,
                                                13, 13, 13, 15, 12, 13, 14, 15 };

            EXPECT_EQ(actual, expected);
        }

        TEST_F(flow_graph, accumulate)
        {
            auto graph = flow_graph_type(grid, { fs::single_flow_router() });

            graph.update_routes(elevation);

            xt::xarray<double> acc = xt::empty_like(elevation);
            xt::xtensor<double, 2> src = xt::ones_like(elevation);
            xt::xtensor<double, 2> expected{
                { 1., 1., 1., 1. }, { 2., 2., 2., 2. }, { 3., 3., 3., 3. }, { 1., 10., 1., 4. }
            };

            graph.accumulate(acc, 1.);
            EXPECT_TRUE(xt::allclose(expected, acc));

            graph.accumulate(acc, src);
            EXPECT_TRUE(xt::allclose(expected, acc));

            EXPECT_TRUE(xt::allclose(expected, graph.accumulate(src)));
            EXPECT_TRUE(xt::allclose(expected, graph.accumulate(1.)));
        }

        TEST_F(flow_graph, basins)
        {
            auto graph = flow_graph_type(grid, { fs::single_flow_router() });

            graph.update_routes(elevation);

            auto actual = graph.basins();

            xt::xtensor<size_type, 2> expected{
                { 1, 1, 1, 3 }, { 1, 1, 1, 3 }, { 1, 1, 1, 3 }, { 0, 1, 2, 3 }
            };

            EXPECT_EQ(actual, expected);

            EXPECT_TRUE(xt::all(xt::equal(xt::flatten(actual), graph.impl().basins())));

            {
                SCOPED_TRACE("with mask");

                xt::xtensor<bool, 2> mask{ { false, false, false, true },
                                           { false, false, false, true },
                                           { false, false, false, true },
                                           { false, false, false, true } };

                graph.set_mask(mask);
                graph.update_routes(elevation);

                auto actual = graph.basins();

                size_type no_basin = std::numeric_limits<size_type>::max();

                xt::xtensor<size_type, 2> expected{ { 1, 1, 1, no_basin },
                                                    { 1, 1, 1, no_basin },
                                                    { 1, 1, 1, no_basin },
                                                    { 0, 1, 2, no_basin } };

                EXPECT_EQ(actual, expected);
            }
        }
    }
}
